Automattic/wp-calypso
# Checkout: Convert product variation picker price calculators to TypeScript

#### Proposed Changes

For some products, checkout displays a variant picker in its first step which allows changing the product's term length in the cart (eg: from yearly to monthly). Unlike the cart items themselves, these variants do not come from the shopping-cart endpoint and so do not benefit from being provided with their prices directly as a localized string. Instead, they calculate their prices on-the-fly using on a fairly complex process.

In order to make that process less error-prone, over time I've converted a number of parts of the product variant system to TypeScript so that it's easier to follow the data that flows through it. However, a recent fatal error (`s.filter(...)[0].minimum_price is undefined`) pointed to a new possible bug in one of the unconverted functions that calculates the price.

In this diff, I convert the functions which create the product variant data (`computeProductsWithPrices()`, `planSlugToPlanProduct()`, and `computeFullAndMonthlyPricesForPlan()`) to TypeScript. Doing this revealed a surprising number of subtle bugs (for example [this code](https://github.com/Automattic/wp-calypso/blob/18171c4d5b9e44cb079ee1455e4596748a101b29/client/state/products-list/selectors/plan-slug-to-plan-product.js#L18-L19) which explicitly passes an object to a function requiring a string, added in https://github.com/Automattic/wp-calypso/pull/57035) which I've also attempted to fix. See comments inline where I marked each one.

> **Note**
> In general, checkout tries not to perform any price math on the client because the server is much better at avoiding floating point errors and performing the appropriate localization for different currencies. The product variant picker, unfortunately, does perform price math, and this diff does not alter that logic. I worry that some of that code is not as correct in some locales as it could be. In particular, the following mutation (from https://github.com/Automattic/wp-calypso/pull/63640) of integer prices doesn't seem like it would hold up to currencies without decimals (eg: JPY). Likely this sort of thing should be addressed in a future diff but I mention it here because I had to touch the code in this diff.

https://github.com/Automattic/wp-calypso/blob/18171c4d5b9e44cb079ee1455e4596748a101b29/client/state/products-list/selectors/compute-full-and-monthly-prices-for-plan.js#L83-L89

#### Testing Instructions

This diff should not alter any behavior except to correct bugs that (presumably) were rarely encountered.

These functions are already covered by unit tests which was extremely helpful as I went back and forth to determine the proper types and behavior. I did not change any of these tests, so at least what they cover should be functionally correct.

For manual testing, there are two cases, Jetpack sites and non-Jetpack sites. The screenshots below use USD as a currency, but a better test would be with your account set to a non-USD currency.

##### Jetpack sites

- Using a Jetpack site, visit checkout with a product in your cart, making sure that product has variants. For example, Jetpack Security can be added using a URL like the following, replacing `example.com` with your own site: http://calypso.localhost:3000/checkout/example.com/jetpack_security_t1_yearly (this is assuming you are running calypso locally; use the appropriate host in the URL if you are using calypso.live).
- Open another window or tab with the same URL except in production wordpress.com. You can skip the product identifier if you wish since it has already been added. For example, replacing `example.com` with your own site: https://wordpress.com/checkout/example.com
- Verify that the variant picker in the first step contains the same prices in both this branch and in production.

<img width="567" alt="jetpack-variant-picker" src="https://user-images.githubusercontent.com/2036909/175426908-4b5152ef-e33e-4334-8602-3013741fb2ac.png">

##### WPCom sites

- Using a non-Jetpack site, visit checkout with a product in your cart, making sure that product has variants. For example, a personal plan can be added using a URL like the following, replacing `example.com` with your own site: http://calypso.localhost:3000/checkout/example.com/personal (this is assuming you are running calypso locally; use the appropriate host in the URL if you are using calypso.live).
- Open another window or tab with the same URL except in production wordpress.com. You can skip the product identifier if you wish since it has already been added. For example, replacing `example.com` with your own site: https://wordpress.com/checkout/example.com
- Verify that the variant picker in the first step contains the same prices in both this branch and in production. You'll have to click "Edit" on the first step to make the variant picker visible.

<img width="566" alt="wpcom-variant-picker" src="https://user-images.githubusercontent.com/2036909/175427140-fcf9af51-7212-4f36-b9d9-a69f029a740a.png">
